include/psa/crypto_struct.h

```diff
@@ -223,6 +223,29 @@ static inline struct psa_key_derivation_s psa_key_derivation_operation_init(
     return v;
 }
 
+struct psa_key_generation_method_s {
```

Comment:
`generation` is not a good name because this type also serves for key derivation. But `derivation` would be equally bad.

`creation` doesn't feel right because that encompasses import and copying, which don't create new key material but do create a key object.

Any other suggestions?

Reply:
You could use `initiation` or `origination` but I am not sure it is really that much better.

Reply:
Perhaps `psa_key_construction_method_t` or perhaps `psa_key_construction_params_t` (it is not just signalling the method) - though I do prefer brevity when it is possible.

Reply:
Maybe:

* `psa_key_spawning_method_t` - _Ok - that was a joke. mostly_
* `psa_key_production_method_t` or `psa_key_production_params_t`?

Reply:
Since the current name is domain_parameters, why not psa_key_domain_parameters_s?
That would also make clear that we're moving that from attributes to an external struct. Wdyt?

Reply:
“Domain parameters” is reasonably established terminology in cryptography for numbers and other mathematical objects that you can choose among a large space in a cryptosystem. A typical example is Diffie-Hellman groups, where the parameters are two primes and a third number, with some required relations between the three). Another example is the various parameters that characterize an elliptic curve; for that, in practice, everyone uses one of a few dozen standard parameters, so we can encode curves as a small integer.

The RSA public exponent technically fits this definition, but it's a bit of a stretch. (Aside: when you consider how RSA works, in fact, it's a better fit to say that the _modulus_ is a domain parameter, and then the public/private exponents are the public/private key. RSA is unusual though in that you can't reuse the modulus: if two keys have the same modulus then (under very mild assumptions) knowing one private key gives you the other key. So in practice, we consider the modulus and the public exponent together to be the public key, and we say that RSA operates in a single domain.)

Other intended usage for this parameter does not fit this mold at all. For example, choosing the prime selection mechanism in RSA, or choosing how inputs are processed in cooked key derivation, don't fit.

In particular, if you need to work with a key, then you need to know its domain parameters. So there should be a way to query a key's domain parameters. But that's precisely something we want to avoid with RSA: the public exponent is already part of the public key, and we don't want to burden attributes with it. And for many other intended use cases, it's impossible to tell from the key which generation method was used.

Reply:
Among the proposals so far, my preference is for “production” as the first noun. It isn't ideal because “production” is vague, and it makes me think of producing the key as in “here's the key” rather than producing the key as in building it. But I like it better than “construction” which makes me think of something you can observe by analyzing the result. Also better than “initiation” or “origination” which make me think they're about _where_ the key material comes from (so importing would be a possibility) rather than _how_ the key material is constructed. And better than “generation” which excludes derivation.

Regarding “method” vs “params” vs “parameters”: to me “method” implies completeness, “parameters” doesn't. On the one hand the value encodes all the information that the generation/derivation process needs to know, so “method” is correct and “parameters” is more vague. On the other hand, these are just some possible ways of tweaking the generation/derivation process, and choosing a specific value seems more specific than what “method” suggests, so “parameters” seems more accurate. I prefer to avoid abbreviations unless they save a lot, so I don't like “params”.

Hence I propose to change the type name to **`psa_key_production_parameters_t`**, and `params` and `params_data_length` for the variable names.

Reply:
Renaming done. (Please wait for the CI to pass before re-reviewing, I may have made mistakes that are only apparent in some configurations, in which case I'll amend the existing commit.)

Having done the rewriting, I'm not so happy about “parameters” being plural. It sometimes makes the documentation ambiguous or awkward, because you can't easily distinguish “parameters” meaning one specific `psa_key_production_parameters_t` structure from “parameters” meaning a set of parameters passed to a function which might include a pointer to such a structure.

Opinions sought: given the resulting wording, especially in [the documentation of the new ext functions](https://github.com/Mbed-TLS/mbedtls/blob/8277ca379e4ac47cbbbf2f858fd002112c5504db/include/psa/crypto.h#L3762), do you prefer “method” or “parameters” (or some alternative proposal)?